<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>WGS BWA pipeline &mdash; gencdna 0.1 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/custom.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="_static/jquery.js?v=5d32c60e"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="_static/documentation_options.js?v=2709fde1"></script>
        <script src="_static/doctools.js?v=888ff710"></script>
        <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Download and extract exons for all human genes" href="all_genes_exons.html" />
    <link rel="prev" title="gencDNA finder for PacBio CCS DNA sequencing" href="index.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            gencdna
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">WGS BWA pipeline</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#circular-consensus">Circular consensus</a></li>
<li class="toctree-l2"><a class="reference internal" href="#sample-fastq-to-fasta">Sample FASTQ to FASTA</a></li>
<li class="toctree-l2"><a class="reference internal" href="#sample-genome-index">Sample genome index</a></li>
<li class="toctree-l2"><a class="reference internal" href="#exon-alignment">Exon alignment</a></li>
<li class="toctree-l2"><a class="reference internal" href="#filtering-unmapped-exons-from-sam">Filtering unmapped exons from SAM</a></li>
<li class="toctree-l2"><a class="reference internal" href="#alignment-coordinates-from-sam">Alignment coordinates from SAM</a></li>
<li class="toctree-l2"><a class="reference internal" href="#find-exon-joins">Find exon joins</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="all_genes_exons.html">Download and extract exons for all human genes</a></li>
<li class="toctree-l1"><a class="reference internal" href="next_steps.html">Next steps</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Advanced:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="synthetic_reads.html">Testing gencDNA finding using synthetic reads</a></li>
<li class="toctree-l1"><a class="reference internal" href="package_structure.html">Python library structure</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Deprecated:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="exons.html">Obtaining exons from individual genes using UCSC Genome Browser</a></li>
<li class="toctree-l1"><a class="reference internal" href="wgs_bowtie.html">Deprecated: WGS pipeline using bowtie2</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">gencdna</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">WGS BWA pipeline</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/wgs_bwa.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="wgs-bwa-pipeline">
<span id="id1"></span><h1>WGS BWA pipeline<a class="headerlink" href="#wgs-bwa-pipeline" title="Link to this heading"></a></h1>
<p>The goal of this pipeline is to find gencDNAs among WGS reads. We look for
gencDNAs by aligning exons to reads then checking if the gap between two
subsequent exons is zero.</p>
<p>This pipeline proceeds by constructing a “genome” out of all reads in a sample,
then indexing this read-genome. With <code class="docutils literal notranslate"><span class="pre">bwa</span> <span class="pre">index</span></code> running single-threaded
(no idea how to multithread this?), this takes about 10 hours per sample.
However, multiple samples can be run in parallel.</p>
<p>Originally, I used Bowtie2, but then switched to BWA since the Bowtie2 pipeline
did not work for all genes. Some exons caused bowtie to search for the
alignment for very long time (I do not know why) and it never seemed to finish.
BWA MEM does not have this issue.</p>
<section id="circular-consensus">
<h2>Circular consensus<a class="headerlink" href="#circular-consensus" title="Link to this heading"></a></h2>
<p>The files I start with are BAM files that contain subreads which need to be
processed using a circular consensus tool <code class="docutils literal notranslate"><span class="pre">ccs</span></code>. I used the default
parameters, which are:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">--min-passes</span></code>: 3</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--min-snr</span></code>: 2.5</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--top-passes</span></code>: 60</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--min-rq</span></code>: 0.99 (min. predicted accuracy)</p></li>
</ul>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>ccs<span class="w"> </span>subreads.bam<span class="w"> </span>reads.fastq
</pre></div>
</div>
<p>This is the only QC - because we want to try not to miss any rare reads that
may contain gencDNA.</p>
</section>
<section id="sample-fastq-to-fasta">
<h2>Sample FASTQ to FASTA<a class="headerlink" href="#sample-fastq-to-fasta" title="Link to this heading"></a></h2>
<p>To create an BWA index, we first need to convert FASTQ to a FASTA file. For
this I made my own Python script:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python<span class="w"> </span>gencdna/file_api/fastq_to_fasta.py<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>reads.fastq<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>reads.fasta
</pre></div>
</div>
</section>
<section id="sample-genome-index">
<h2>Sample genome index<a class="headerlink" href="#sample-genome-index" title="Link to this heading"></a></h2>
<p>When supplied with a FASTA file containing multiple sequences, BWA does
concatenation behind the scenes (and parses everything back), so we can build
an index directly from FASTA file form pre-processed reads:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>bwa<span class="w"> </span>index<span class="w"> </span>-p<span class="w"> </span>&lt;read_genome_prefix&gt;<span class="w"> </span>reads.fasta
</pre></div>
</div>
<p>This is helpful, since I previously created code to do this manually with
bowtie2.</p>
</section>
<section id="exon-alignment">
<h2>Exon alignment<a class="headerlink" href="#exon-alignment" title="Link to this heading"></a></h2>
<p>I used the default BWA MEM parameters for exon alignment:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>bwa<span class="w"> </span>mem<span class="w"> </span>-t<span class="w"> </span>&lt;num_cores&gt;<span class="w"> </span>-a<span class="w"> </span>&lt;read_genome_prefix&gt;<span class="w"> </span>&lt;exons_fasta&gt;<span class="w"> </span>&gt;<span class="w"> </span>&lt;exons_sam&gt;
</pre></div>
</div>
</section>
<section id="filtering-unmapped-exons-from-sam">
<h2>Filtering unmapped exons from SAM<a class="headerlink" href="#filtering-unmapped-exons-from-sam" title="Link to this heading"></a></h2>
<p>Unlike in bowtie2, there does not seem to be an option in <code class="docutils literal notranslate"><span class="pre">bwa</span></code> to omit
storing exons in the alignment SAM file that did not match any read. Therefore,
I added this step to filter those exons out:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>samtools<span class="w"> </span>view<span class="w"> </span>-F<span class="w"> </span><span class="m">4</span><span class="w"> </span>-o<span class="w"> </span>&lt;mapped_exons_sam&gt;<span class="w"> </span>&lt;exons_sam&gt;
</pre></div>
</div>
<p><code class="docutils literal notranslate"><span class="pre">-F</span></code> flag filters out all reads that do not match 4 in the flag field. In bitwise
representation 4 represents unmapped reads. See page 7 at <a class="reference external" href="https://samtools.github.io/hts-specs/SAMv1.pdf">https://samtools.github.io/hts-specs/SAMv1.pdf</a>
for details.</p>
</section>
<section id="alignment-coordinates-from-sam">
<h2>Alignment coordinates from SAM<a class="headerlink" href="#alignment-coordinates-from-sam" title="Link to this heading"></a></h2>
<p>Now that SAM file is free from exons that did not align to any read, we use
Python script to obtain alignment coordinates for each exon on each read:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python<span class="w"> </span>gencdna/file_api/alignment_coords.py<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>&lt;mapped_exons_sam&gt;<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>&lt;mapped_exons_coords_csv&gt;
</pre></div>
</div>
</section>
<section id="find-exon-joins">
<h2>Find exon joins<a class="headerlink" href="#find-exon-joins" title="Link to this heading"></a></h2>
<p>This script calculates gaps between adjacent exons then keeps the adjacent
reads where that gap is zero:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python<span class="w"> </span>gencdna/file_api/filter_gapped_exon_alignments.py<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>&lt;mapped_exons_coords_csv&gt;<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>&lt;mapped_exons_joins_csv&gt;
</pre></div>
</div>
<p>This file will list each exon for each read in a single row that has a 0 gap.</p>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="index.html" class="btn btn-neutral float-left" title="gencDNA finder for PacBio CCS DNA sequencing" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="all_genes_exons.html" class="btn btn-neutral float-right" title="Download and extract exons for all human genes" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, Igor Segota.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>